### PR TITLE
Added read catalog example for iOS and Android

### PR DESCRIPTION
**Pull request by borzun**
- configured read example for iOS and Android platfroms;
- refactored CMakeLists.txt for examples generation in order to reduce code duplication;
- fixed problem with dataservice-write-example library building on Android.

Resolves: OLPEDGE-287, OLPEDGE-288

Signed-off-by: Bohdan Kurylovych <ext-bohdan.kurylovych@here.com>

### CODE REVIEW

File: examples/dataservice-read/CMakeLists.txt [hunk 1]
```diff
@@ -16,16 +16,39 @@
...
+                                ${EDGE_SDK_EXAMPLE_SUCCESS_STRING}
+                                ${EDGE_SDK_EXAMPLE_FAILURE_STRING})
 
-project(dataservice-read-example)
+elseif (ANDROID)
+    include(${CMAKE_CURRENT_SOURCE_DIR}/../helpers/cmake/gen_android_example.cmake.in)
 
-add_executable(dataservice-read-example example.cpp)
+    set(EDGE_SDK_EXAMPLE_PACKAGE_NAME "dataservice_read_example")
+    set(EDGE_SDK_EXAMPLE_PACKAGE_JNI_NAME "dataservice_1read_1example")
```
Comment (mykhailo-kuchma) [inline examples/dataservice-read/CMakeLists.txt]: is it done intentionally?

Comment (borzun) [reply examples/dataservice-read/CMakeLists.txt]: You meant about 1s? Yes, it is for JNI - basically, it is how JNI 'translate' JAVA native methods.

Alternatively, we can invoke JNI compiler to generate those methods' names automatically during CMakeLists.txt execution and replace generated names via configure_file(). It is more complex way, but, I guess, it is how it should be done.

File: docs/dataservice-read-catalog-example.md [hunk 2]
```diff
@@ -32,6 +32,81 @@ To execute the example, run:
 
 If everything is fine, you would see `hrn:here:data:::here-optimized-map-for-visualization-2` catalog description, number of partitions and the size of `100000093` partition.
 
+## Building and running on Android
+
+This example shows how to integrate the HERE OLP Edge SDK C++ libraries into the Android project and shows how to read partitions' data from the catalog.
```
Comment (andescu) [inline docs/dataservice-read-catalog-example.md]:
```suggestion
This example shows how to integrate the HERE OLP Edge SDK C++ libraries into the Android project and shows how to read data from a catalog layer and partition.
```

File: docs/dataservice-read-catalog-example.md [hunk 3]
```diff
@@ -32,6 +32,81 @@ To execute the example, run:
...
+
+### Prerequisites
+
+* Setup the Android environment.
+* Provide correct keyId and secret key in the `example.cpp` (refer to the beginning of this README file).
```
Comment (andescu) [inline docs/dataservice-read-catalog-example.md]:
```suggestion
* Provide correct keyId and secret key in the `example.cpp` (refer to the beginning of this file).
```

File: docs/dataservice-read-catalog-example.md [hunk 4]
```diff
@@ -9,7 +9,7 @@ const std::string gKeyId(""); // your here.access.key.id
 const std::string gKeySecret(""); // your here.access.key.secret
```
Comment (andescu) [inline docs/dataservice-read-catalog-example.md]: Please mark for next commit that we need to align these with the coding style.
https://google.github.io/styleguide/cppguide.html#Constant_Names

File: docs/dataservice-write-example.md [hunk 5]
```diff
@@ -105,7 +105,7 @@ Now open the generated `XCode` project:
 open olp-cpp-sdk.xcodeproj
 ```
 
-Select the `dataservice-write-example` scheme from the schemes list in `XCode` and specify your signing credentials for the `dataservice-write-example` target.
+Select the `dataservice-write-example` scheme from the schemes list in `XCode` project and specify your signing credentials for the `dataservice-write-example` target.
```
Comment (andescu) [inline docs/dataservice-write-example.md]:
```suggestion
Select the `dataservice-write-example` scheme from the schemes list in the `XCode` project and specify your signing credentials for the `dataservice-write-example` target.
```

File: examples/dataservice-read/CMakeLists.txt [hunk 6]
```diff
@@ -16,16 +16,39 @@
 # License-Filename: LICENSE
 cmake_minimum_required(VERSION 3.5)
 
-if (IOS OR ANDROID)
-    message(WARNING "Unsupported architecture!")
-    return()
-endif()
+set(EDGE_SDK_DATASERVICE_READ_EXAMPLE_TARGET dataservice-read-example)
+
+set(EDGE_SDK_EXAMPLE_SUCCESS_STRING "Reading partition's data from catalog performed successfully")
```
Comment (andescu) [inline examples/dataservice-read/CMakeLists.txt]:
```suggestion
set(EDGE_SDK_EXAMPLE_SUCCESS_STRING "Reading the partition data from the catalog completed successfully")
```

File: examples/dataservice-read/CMakeLists.txt [hunk 7]
```diff
@@ -16,16 +16,39 @@
...
+set(EDGE_SDK_EXAMPLE_FAILURE_STRING "Failed to read data from the catalog!")
```
Comment (andescu) [inline examples/dataservice-read/CMakeLists.txt]:
```suggestion
set(EDGE_SDK_EXAMPLE_FAILURE_STRING "Failed to read data from the specified catalog!")
```

File: examples/dataservice-read/CMakeLists.txt [hunk 8]
```diff
@@ -16,16 +16,39 @@
...
+
+if (IOS)
+    include(${CMAKE_CURRENT_SOURCE_DIR}/../helpers/cmake/gen_ios_example.cmake.in)
+
```
Comment (andescu) [inline examples/dataservice-read/CMakeLists.txt]: remove one newline

File: examples/dataservice-read/CMakeLists.txt [hunk 9]
```diff
@@ -16,16 +16,39 @@
...
+
+    gen_ios_example_application(${EDGE_SDK_DATASERVICE_READ_EXAMPLE_TARGET}
+                                ${EDGE_SDK_EXAMPLE_SUCCESS_STRING}
+                                ${EDGE_SDK_EXAMPLE_FAILURE_STRING})
 
-project(dataservice-read-example)
+elseif (ANDROID)
+    include(${CMAKE_CURRENT_SOURCE_DIR}/../helpers/cmake/gen_android_example.cmake.in)
```
Comment (andescu) [inline examples/dataservice-read/CMakeLists.txt]: can we not use ${CMAKE_SOURCE_DIR/examples/helpers/....

File: examples/dataservice-read/example.cpp [hunk 10]
```diff
@@ -164,6 +166,7 @@ int main() {
     LOG_WARNING(kLogTag, "Request partition metadata is not present!");
   }
 
+  int returnCode = -1;
```
Comment (andescu) [inline examples/dataservice-read/example.cpp]:
```suggestion
  int return_code = -1;
```

File: examples/dataservice-write/CMakeLists.txt [hunk 11]
```diff
@@ -17,112 +17,27 @@
...
-    # Configure CMakeLists.txt.in with specified variables
-    set(EDGE_SDK_EXAMPLE_TARGET_NAME ${EDGE_SDK_DATASERVICE_WRITE_EXAMPLE_TARGET})
-    set(EDGE_SDK_SUCCESSFUL_EXAMPLE_EXECUTION_STRING "Publish Successfull")
-    set(EDGE_SDK_FAILING_EXAMPLE_EXECUTION_STRING "Publish Failed!")
-    configure_file(
-        ${CMAKE_CURRENT_SOURCE_DIR}/../helpers/ios/CMakeLists.txt.in
-        ${CMAKE_CURRENT_BINARY_DIR}/ios/CMakeLists.txt
-        @ONLY
-    )
+    include(${CMAKE_CURRENT_SOURCE_DIR}/../helpers/cmake/gen_ios_example.cmake.in)
```
Comment (andescu) [inline examples/dataservice-write/CMakeLists.txt]: same. Use PROJECT_SOURCE_DIR?

File: docs/dataservice-read-catalog-example.md (hunk 4 above)
Comment (borzun) [reply docs/dataservice-read-catalog-example.md]: I changed the naming of global and local variables according to the C++ coding style. Please, check both examples.cpp

**Pull request merged**